keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -413,28 +412,83 @@ On Pod Create/Update/Delete:
 
 ### Kube-Proxy
 
-Kube-proxy will be modified to consume EndpointSlice instances besides Endpoints resource. A flag will be added to kube-proxy to toggle the mode. 
-
-```
-Watch: Service, EndpointSlice ==> Manage: iptables, ipvs, etc
-```
-- Merge multiple EndpointSlice into an aggregated list.
-- Reuse the existing processing logic 
-
-### Endpoint Controller (classic)
-
-In order to ensure backward compatibility for external consumer of the core/v1 Endpoints API, the existing K8s endpoint controller will keep running until the API is EOL. The following limitations will apply:
-- Starting from EndpointSlice beta: If # of endpoints in one Endpoints object exceed 1000, generate a warning event to the object. 
-- Starting from EndpointSlice GA: Only include up to 1000 endpoints in one Endpoints Object and throw events.
+Kube-Proxy will support consuming EndpointSlice resources as an alternative to
+Endpoints resources. This will be enabled with the `EndpointSliceProxying`
+feature gate.
+
+This will require minimal changes to proxier implementations, instead updating
+the shared code that transforms Endpoints into a generic data structure to also
+transform EndpointSlices into that same data structure.
+
+### Endpoints Controller (classic)
+
+In order to ensure backwards compatibility for external consumer of the core/v1
+Endpoints API, the existing Endpoints controller will keep running until the
+API is EOL. As EndpointSlices become more stable, the Endpoints controller will
+gradually limit functionality.
+
+* Kubernetes 1.19: If the number of endpoints in one Endpoints object exceeds
+  1000, generated a warning event on the Endpoints resource.
+* Kubernetes 1.20: Limit the number of endpoints in one Endpoints object to
+  1000 and generate warning events on Endpoints resource if additional endpoints
```

Comment:
how was this limit selected? Does this map to a specific problematic quantity or threshold? How frequently would events be created to indicate this?

Reply:
That's a great question, @freehan probably has more background on this. I'm not tied to this in any way, it's something from the original KEP that I've clarified here, but I'm happy to push the timeline on this and/or change the parameters.

Reply:
Scalability team has a [recommended threshold](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md) based on testing. For endpoint, the threshold is 250. Above 250 endpoints with many services and nodes, API server starts to get stressed. 

In reality, if a cluster has one or few service with 1000+ endpoints that rarely changes, it does not hurt that much. Hence, picking the threshold and warning.

Reply:
putting an Event creation in the Endpoint controller sync path could be impactful depending on how often this was written. was an annotation on the Endpoints object considered instead?

Reply:
I had originally thought that an event would be more visible, especially if it was a warning. I'd imagine that watching a cluster for warning events is reasonably common. With that said, I'd already thought we'd have to have some kind of rate-limiting mechanism here to only publish an event like this intermittently for each Endpoints resource.

As I've thought about it more, I like your annotation idea. It seems like a warning event might actually be too noisy for something that users can't actually do anything about (other than disabling the Endpoints controller or having smaller Services, neither of which we want). 

To go one step further, would a label be appropriate here? I think it could be useful to easily select all Endpoints in a cluster that are currently over-capacity to understand the impact of upgrading to the release that will truncate those Endpoints resources. Maybe a `endpoints.kubernetes.io/over-capacity: "true"` label? This could persist after the release where Endpoints start to get truncated as a helpful indication of the effect of this change.

Reply:
Pushed an update with this approach. Added a slight tweak to it that may not make sense. In 1.20 `endpoints.kubernetes.io/over-capacity` will be set to `warning` and in 1.21 it will be set to `truncated`.